extractor.go

```diff
@@ -52,30 +82,66 @@ type File struct {
 }
 
 type Extractor struct {
-	enc      *json.Encoder
-	num      int // TODO(dennwc): use it!
+	mu  sync.Mutex
+	enc *json.Encoder
+
+	num      int
 	maxSize  int64
 	symLinks bool
 }
 
+type extractJob struct {
+	fname string
+	path  string
+	buf   []byte // sample buffer
+}
+
 // Extract imports recursively from a given directory. The root is a root of the project's repository and rel is the
 // relative path inside it that will be processed. Two paths exists to allow the library to potentially parse dependency
 // manifest files that are usually located in the root of the project.
 func (e *Extractor) Extract(root, rel string) error {
+	var (
+		jobs      chan *extractJob
+		errc      chan error
+		wg        sync.WaitGroup
+		sampleBuf []byte
+	)
+	if e.num != 1 {
+		jobs = make(chan *extractJob, e.num)
+		errc = make(chan error, e.num)
+		for i := 0; i < e.num; i++ {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				e.worker(jobs, errc)
+			}()
+		}
+		// no sample buffer, it's per-routine
```

Comment:
Lets move it to some function. 
Btw. I think error channel is not closed.

Reply:
Sorry, but I don't think it will make the code smaller.

If you propose to move the worker creation, it will require to pass `wg` to a function, which is considered a bad practice (you should see all usages of `wg` in one place, if possible).

Same for other things: creating a function will require passing other state like channels to them. Having in mind the size of the code inlining it here may still be acceptable.

But if you have some specific suggestion, please elaborate a bit more.

Re: not closing the channel - it's not necessary. GC will take care of that anyway. It's not a resource leak.

Reply:
my intention was not make the code shorter but more readable (less statements from different contexts, like here we start workers, there we handle channels plus *ifs* single threaded or multi threaded.

Reply:
As I said, if you have any suggestions - please elaborate. Moving things will complicate code or will require passing all the state as arguments.